### PR TITLE
SI-7088 Array crasher in erasure.

### PR DESCRIPTION
**Pull request by paulp**
The usual business where half our pattern matches are missing half the
necessary cases.

### CODE REVIEW

**Comment by paulp** [comment]
Review by @retronym

File: src/reflect/scala/reflect/internal/transform/Erasure.scala [hunk 1]
```diff
@@ -55,9 +55,13 @@ trait Erasure {
     }
   }
 
+  /** Arrays despite their finality may turn up as refined type parents,
+   *  e.g. with "tagged types" like Array[Int] with T.
+   */
   protected def unboundedGenericArrayLevel(tp: Type): Int = tp match {
     case GenericArray(level, core) if !(core <:< AnyRefClass.tpe) => level
-    case _ => 0
+    case RefinedType(ps, _) if ps.nonEmpty                        => logResult(s"Unbounded generic level for $tp is")(ps map unboundedGenericArrayLevel max)
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala]: I feel this ought to be done inside the `GenericArray` extractor. A few more tests to show the interaction with nested arrays and existentials might flush this out.

``` scala
(Array[T] with Tag) forSome { type T }
Array[Array[T] with Tag]
```

You might also change the comment to `Array[String] with T`, as this particular code is geared towards generic arrays. (This also points to more useful test cases involving primitive arrays.)

Comment (paulp) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala]: This is what I get for snipping the minimal patch off the big patch...

My actual patch for this subject matter is 

```
 src/compiler/scala/tools/nsc/backend/ScalaPrimitives.scala   |   8 +-
 src/compiler/scala/tools/nsc/transform/AddInterfaces.scala   |   2 +-
 src/compiler/scala/tools/nsc/transform/Erasure.scala         |  95 ++++++------
 src/compiler/scala/tools/nsc/transform/Flatten.scala         |   2 +-
 src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala |  21 +--
 src/compiler/scala/tools/nsc/transform/UnCurry.scala         |  40 +++--
 src/compiler/scala/tools/nsc/typechecker/Implicits.scala     |   2 +-
 src/reflect/scala/reflect/internal/Definitions.scala         |  77 ++++++++--
 src/reflect/scala/reflect/internal/Types.scala               |  11 +-
 src/reflect/scala/reflect/internal/tpe/GlbLubs.scala         |  16 ++
 src/reflect/scala/reflect/internal/transform/Arrays.scala    | 103 +++++++++++++
 src/reflect/scala/reflect/internal/transform/Erasure.scala   | 261 ++++++++++++---------------------
 test/files/run/analyzerPlugins.check                         |   4 +-
 test/files/run/existentials.scala                            |  22 +--
 test/files/run/t5353.check                                   |  59 +++++++-
 test/files/run/t5353.scala                                   | 115 ++++++++++++++-
 test/files/run/t5353b.check                                  | 150 +++++++++++++++++++
 test/files/run/t5353b.scala                                  | 357 +++++++++++++++++++++++++++++++++++++++++++++
 18 files changed, 1042 insertions(+), 303 deletions(-)
```

But I didn't want poor #7088 to wait on my breaking that into beautiful little bird appetite sized chunks.

Comment (retronym) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala]: LGTM as the starting point, so long as you open a bug so can keep track of the residuals and you add comment here to mention as much.

**Comment by paulp** [comment]
https://issues.scala-lang.org/browse/SI-7522

**Pull request merged**